plan/typeinferer.go

```diff
@@ -221,6 +221,45 @@ func mergeArithType(fta, ftb *types.FieldType) byte {
 	return mysql.TypeLonglong
 }
 
+func mergeCmpType(fta, ftb *types.FieldType) (ft *types.FieldType) {
+	ft = &types.FieldType{}
+	if fta.Charset == charset.CharsetUTF8 && ftb.Charset == charset.CharsetUTF8 {
+		ft.Charset = charset.CharsetUTF8
+		ft.Collate = "utf8_bin"
```

Comment:
In other cases, ft.Charset/Collate will be set as CharsetBin/CollateBin outside this func.